### DISCUSSION TITLE
How to modify the coefficients of the objective in FG_eval for ipopt?

### DISCUSSION THREAD

**Discussion by type-null** [coin-or/CppAD#180] [Q&A]
I am exploring the ipopt example file https://github.com/coin-or/CppAD/blob/master/example/ipopt_solve/get_started.cpp when I want to solve the problem again with only the coeffient changes in the objective function. I wonder if there is a fast way to accomplish this.

Currently, as written in the example file, the objective is written by the `operator()` function in the `FG_eval` class. Is it possible, by passing the new coefficients to the previous FG_eval, to change the `fg[0]` without changing `fg[1]` and `fg[2]` or rewriting `fg[1], fg[2]`? 

Thanks a lot!

**Comment by bradbell** [comment]
It seems like you want to be able to change a coefficient in the definition of the FG_eval operator() function. You could make this coefficients a member variable in the FG_eval class that is  set by the constructor. 

I am not sure if this would do what you want ?

**Reply by type-null** [reply]
Yes, this is what I want. But I am not sure if the ipopt::solve will call the operator() in FG_eval and write all other constraints all over again.

**Comment by bradbell** [comment]
Each time you call ipopt::solve, it will call the operator() function.
If retape is true, it will call the operator function each time x changes; see
https://cppad.readthedocs.io/en/latest/ipopt_solve.html#retape

**Reply by type-null** [reply]
Could you refer me to some explanation on retape? I found the documentation hard to understand. Besides, the variables x is not changed, only fg[0] needs change.

**Comment by bradbell** [comment]
The function fg_eval is used to 'record' the function x -> fg(x).  The optimizer makes callbacks to ipopt::solve to evaluate fg(x) , and its derivatives, at different values of x. The recording takes a significant amount of time relative to evaluating the evaluations. The retape option handles the case where the recording is different from the function; e.g., if there is an if - then block in the function. See the discussion below operation on
https://cppad.readthedocs.io/en/latest/glossary.html#operation

**Reply by type-null** [reply]
I still could not understand `retape` but I tried a few things. It turned out not allowing me to use the x and fg from the same fg_eval twice: every time I called `ipopt::solve`, it seemed to give a new `x` to the fg_eval that is not allowed to be changed.

Here's an example to explain what I am trying to do:
We first solve this problem using cppad
$$\min x_0^2 \quad s.t. 1 \le x_0 \le 3, 2 \le x_1 \le 4.$$
Next, we want to solve the problem below:
$$\min x_1^2 \quad s.t. 1 \le x_0 \le 3, 2 \le x_1 \le 4.$$
where only the objective has changed. 

In our actual case, the constraints are huge. So it is painful to reconstruct all the constraints again just to solve it with a different objective. And we need to do this (modify only objective and solve) hundreds of times. So I hope to find a way to just change `fg[0]` and reuse everything else that is already constructed to save time.

Thanks a lot!

**Reply by type-null** [reply]
Is it possible to replace the operator() function with a customized function somehow?

**Comment by bradbell** [comment]
The operator function can be any function you define that satisfies the requirements for fg_eval on:
https://cppad.readthedocs.io/en/latest/ipopt_solve.html#fg-eval
I can use any member variables that are defined in you FG_eval class

**Reply by type-null** [reply]
Thanks! Now I am solving these similar models repeatedly, everytime building a new model (fg_eval). But after several hours of running, my machine warned me about **low disk space**. I wonder if CppAD or Ipopt needs to manually delete solved model objects to release space. Appreciate your help!

**Comment by bradbell** [comment]
1. Is the disk space still low after the program exits ?
2. If you look at memory usage is it climbing while the program is executing ?

**Reply by type-null** [reply]
1. Yes, although I used ctrl+C to force quit the program.
2. No, it was stable at a normal level.

**Comment by bradbell** [comment]
Do you know the location / name of the disk files that ate taking up the space ? I suspect they are in /temp or somthing like that.
